metals/src/main/scala/scala/meta/internal/metals/codeactions/CreateCompanionObjectCodeAction.scala

```diff
@@ -0,0 +1,212 @@
+package scala.meta.internal.metals.codeactions
+
+import java.util.Scanner
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.Defn
+import scala.meta.Template
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ServerCommands
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.lsp4j.Location
+import org.eclipse.{lsp4j => l}
+
+/**
+ * look at the
+ *
+ * @param trees
+ */
+class CreateCompanionObjectCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorRewrite
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val uri = params.getTextDocument().getUri()
+
+    val path = uri.toAbsolutePath
+    val range = params.getRange()
+    val applyTree =
+      if (range.getStart == range.getEnd)
+        trees
+          .findLastEnclosingAt[Tree](
+            path,
+            range.getStart(),
+            applyWithSingleFunction
+          )
+      else
+        None
+
+    applyTree.collect { case tree: Tree =>
+      val name = getName(tree)
+
+      findCompanionObject(tree, name) match {
+        case Some(comanionObject) =>
+          buildShowingCompanionObjectCodeAction(comanionObject, uri)
+        case None =>
+          val document = buffers.get(path).getOrElse("")
+          val indentationString =
+            getIndentationForPositionInDocument(tree.pos, document)
+
+          buildCreatingCompanionObjectCodeAction(
+            path,
+            tree,
+            uri,
+            indentationString,
+            name,
+            hasBraces(tree, document)
+          )
+      }
+    }.toSeq
+
+  }
+
+  private def getIndentationForPositionInDocument(
+      treePos: Position,
+      document: String
+  ) = {
+    val scanner = new Scanner(document)
+
+    (0 to treePos.startLine)
+      .map(_ => scanner.nextLine())
+      .last
+      .takeWhile(_.isWhitespace)
+  }
```

Comment:
```suggestion
  private def getIndentationForPositionInDocument(
      treePos: Position,
      document: String
  ) = {
    document.substring(treePos.start - treePos.endColumn, treePos.start).takeWhile(_.isWhitespace)
  }
```

this way we don't have iterate over the entire document.

Reply:
Great! Thank you!